src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -107,644 +93,557 @@ public void Emit()
                     {
                         _typeIndex.Add(spec.TypeRef, spec);
                     }
 
-                    foreach (TypeGenerationSpec typeGenerationSpec in _currentContext.GeneratedTypes)
+                    foreach (TypeGenerationSpec typeGenerationSpec in contextGenerationSpec.GeneratedTypes)
                     {
-                        GenerateTypeInfo(typeGenerationSpec);
+                        SourceText? sourceText = GenerateTypeInfo(contextGenerationSpec, typeGenerationSpec);
+                        if (sourceText != null)
+                        {
+                            _sourceGenerationContext.AddSource($"{contextGenerationSpec.ContextType.Name}.{typeGenerationSpec.TypeInfoPropertyName}.g.cs", sourceText);
+                        }
                     }
 
-                    string contextName = _currentContext.ContextType.Name;
+                    string contextName = contextGenerationSpec.ContextType.Name;
 
                     // Add root context implementation.
-                    AddSource(
-                        $"{contextName}.g.cs",
-                        GetRootJsonContextImplementation(),
-                        isRootContextDef: true);
+                    _sourceGenerationContext.AddSource($"{contextName}.g.cs", GetRootJsonContextImplementation(contextGenerationSpec));
 
                     // Add GetJsonTypeInfo override implementation.
-                    AddSource($"{contextName}.GetJsonTypeInfo.g.cs", GetGetTypeInfoImplementation(contextGenerationSpec), interfaceImplementation: JsonTypeInfoResolverTypeRef);
+                    _sourceGenerationContext.AddSource($"{contextName}.GetJsonTypeInfo.g.cs", GetGetTypeInfoImplementation(contextGenerationSpec));
 
                     // Add property name initialization.
-                    AddSource($"{contextName}.PropertyNames.g.cs", GetPropertyNameInitialization());
+                    _sourceGenerationContext.AddSource($"{contextName}.PropertyNames.g.cs", GetPropertyNameInitialization(contextGenerationSpec));
 
+                    _propertyNames.Clear();
                     _typeIndex.Clear();
                 }
             }
 
-            private void AddSource(string fileName, string source, bool isRootContextDef = false, string? interfaceImplementation = null)
+            private static SourceWriter CreateSourceWriterWithContextHeader(ContextGenerationSpec contextSpec, bool isPrimaryContextSourceFile = false, string? interfaceImplementation = null)
             {
-                string? generatedCodeAttributeSource = isRootContextDef ? s_generatedCodeAttributeSource : null;
+                var writer = new SourceWriter();
 
-                ImmutableEquatableArray<string> declarationList = _currentContext.ContextClassDeclarations;
-                int declarationCount = declarationList.Count;
-                Debug.Assert(declarationCount >= 1);
+                writer.WriteLine("""
```

Comment:
Could this method be renamed given that this isn't one line?

Reply:
Any suggestions? I figured it should follow `StringBuilder` convention in which "Line" suggests that a newline is inserted at the end of the string, which is happening here.

Reply:
If you mean `IndentedSourceWriter`, the type doesn't support multi-line string literals. It's intended as a direct replacement of that type.